aws/session/session.go

```diff
@@ -12,87 +12,56 @@ import (
 
 // NewPoolSafe is used to create a pool of AWS sessions with different profile/region permutations
 func NewPoolSafe(profiles []string, regions []string, logger *log.Logger) (allSessions *PoolSafe) {
+	sessions := map[string]*Pool{}
 
-	wg := sync.WaitGroup{}
-	sp := &PoolSafe{
-		Sessions: make(map[string]*Pool),
-	}
-
-	if len(regions) != 0 {
-		for _, p := range profiles {
-			for _, r := range regions {
-				wg.Add(1)
-				// Wait until we have the session for each permutation of profiles and regions
-				go func(p string, r string) {
-					defer wg.Done()
-
-					s, err := newSession(p, r)
-					if err != nil {
-						logger.Fatalf("Error when trying to create session:\n%v", err)
-					}
+	for _, region := range regions {
+		if stsCredentialsSet() {
+			pool := newPool("", region, logger)
+			poolName := fmt.Sprintf("default-%s", region)
+			sessions[poolName] = pool
 
-					if err = validateSessionCreds(s); err != nil {
-						logger.Fatal(err)
-					}
+			continue
+		}
 
-					session := Pool{
-						Logger:      logger,
-						ProfileName: p,
-						Session:     s,
-					}
-					sp.Sessions[fmt.Sprintf("%s-%s", p, r)] = &session
-				}(p, r)
-			}
+		for _, profile := range profiles {
+			pool := newPool(profile, region, logger)
+			poolName := fmt.Sprintf("%s-%s", profile, region)
+			sessions[poolName] = pool
 		}
-	} else {
-		for _, p := range profiles {
-			wg.Add(1)
-			// Wait until we have the session for each profile
-			go func(p string) {
-				defer wg.Done()
+	}
 
-				s, err := newSession(p, "")
-				if err != nil {
-					logger.Fatalf("Error when trying to create session:\n%v", err)
-				}
+	return &PoolSafe{Sessions: sessions}
+}
 
-				if err = validateSessionCreds(s); err != nil {
-					logger.Fatal(err)
-				}
+func stsCredentialsSet() bool {
+	sessionToken := os.Getenv("AWS_SESSION_TOKEN")
+	accessKeyID := os.Getenv("AWS_ACCESS_KEY_ID")
+	secretAccessKey := os.Getenv("AWS_SECRET_ACCESS_KEY")
 
-				session := Pool{
-					Logger:      logger,
-					ProfileName: p,
-					Session:     s,
-				}
-				sp.Sessions[fmt.Sprintf("%s", p)] = &session
-			}(p)
-		}
+	if sessionToken == "" || accessKeyID == "" || secretAccessKey == "" {
+		return false
 	}
-	// Wait until all sessions have been initialized
-	wg.Wait()
 
-	return sp
+	return true
 }
 
-func validateSessionCreds(session *session.Session) (err error) {
-	creds := session.Config.Credentials
-	if _, err := creds.Get(); err != nil {
-		return fmt.Errorf("Error when validating credentials:\n%v", err)
+func newPool(profile string, region string, logger *log.Logger) *Pool {
```

Comment:
Pool was intended to refer to the pool of session objects, not an individual session. Consider renaming?

Reply:
I just named based on the object it was returning. I agree the naming could be better. Will work on fixing that